crypt.3
.\" Written and revised by Solar Designer <solar at openwall.com> in 2000-2011.
.\" Revised by Zack Weinberg <zackw at panix.com> in 2017.
.\" Converted to mdoc format by Zack Weinberg in 2018.
.\"
.\" No copyright is claimed, and this man page is hereby placed in the public
.\" domain.  In case this attempt to disclaim copyright and place the man page
.\" in the public domain is deemed null and void, then the man page is
.\" Copyright 2000-2011 Solar Designer, 2017, 2018 Zack Weinberg, and it is
.\" hereby released to the general public under the following terms:
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted.
.\"
.\" There's ABSOLUTELY NO WARRANTY, express or implied.
.\"
.Dd October 11, 2017
.Dt CRYPT 3
.Os "Openwall Project"
.Sh NAME
.Nm crypt , crypt_r , crypt_rn , crypt_ra
.Nd passphrase hashing
.Sh LIBRARY
.Lb libcrypt
.Sh SYNOPSIS
.In crypt.h
.Ft "char *"
.Fo crypt
.Fa "const char *phrase"
.Fa "const char *setting"
.Fc
.Ft "char *"
.Fo crypt_r
.Fa "const char *phrase"
.Fa "const char *setting"
.Fa "struct crypt_data *data"
.Fc
.Ft "char *"
.Fo crypt_rn
.Fa "const char *phrase"
.Fa "const char *setting"
.Fa "struct crypt_data *data"
.Fa "int size"
.Fc
.Ft "char *"
.Fo crypt_ra
.Fa "const char *phrase"
.Fa "const char *setting"
.Fa "void **data"
.Fa "int *size"
.Fc
.Sh DESCRIPTION
The
.Nm crypt ,
.Nm crypt_r ,
.Nm crypt_rn ,
and
.Nm crypt_ra
functions irreversibly
.Dq hash
.Fa phrase
for storage in the system password database
.Pq Xr shadow 5
using a cryptographic
.Dq hashing method.
The result of this operation is called a
.Dq hashed passphrase
or just a
.Dq hash.
Hashing methods are described in
.Xr crypt 5 .
.Pp
.Fa setting
controls which hashing method to use,
and also supplies various parameters to the chosen method,
most importantly a random
.Dq salt
which ensures that no two stored hashes are the same,
even if the
.Fa phrase
strings are the same.
.Pp
The
.Fa data
argument to
.Nm crypt_r
is a structure of type
.Vt "struct crypt_data" .
It has at least these fields:
.Bd -literal -offset indent
struct crypt_data {
    char output[CRYPT_OUTPUT_SIZE];
    char setting[CRYPT_OUTPUT_SIZE];
    char phrase[CRYPT_MAX_PASSPHRASE_SIZE];
    char initialized;
};
.Ed
.Pp
Upon a successful return from
.Nm crypt_r ,
the hashed passphrase will be stored in
.Fa output .
Applications are encouraged, but not required, to use the
.Fa phrase
and
.Fa setting
fields to store the strings that they will pass as
.Fa phrase
and
.Fa setting
to
.Nm crypt_r .
This will make it easier to erase all sensitive data
after it is no longer needed.
.Pp
The
.Fa initialized
field must be set to zero before the first time a
.Vt "struct crypt_data"
object is first used in a call to
.Fn crypt_r .
We recommend zeroing the entire object,
not just
.Fa initialized
and not just the documented fields,
before the first use.
(Of course, do this before storing anything in
.Fa setting
and
.Fa phrase . )
.Pp
The
.Fa data
argument to
.Nm crypt_rn
should also point to a
.Vt "struct crypt_data"
object, and
.Fa size
should be the size of that object, cast to
.Vt int .
When used with
.Nm crypt_rn ,
the entire
.Fa data
object (except for the
.Fa phrase
and
.Fa setting
fields) must be zeroed before its first use;
this is not just a recommendation, as it is for
.Nm crypt_r .
Otherwise, the fields of the object have the same uses that they do for
.Nm crypt_r .
.Pp
On the first call to
.Nm crypt_ra ,
.Fa data
should be the address of a
.Vt "void *"
variable set to NULL, and
.Fa size
should be the address of an
.Vt int
variable set to zero.
.Nm crypt_ra
will allocate and initialize a
.Vt "struct crypt_data"
object, using
.Xr malloc 3 ,
and write its address and size into the variables pointed to by
.Fa data
and
.Fa size .
These can be reused in subsequent calls.
After the application is done hashing passphrases,
it should deallocate the
.Vt "struct crypt_data"
object using
.Xr free 3 .
.Sh RETURN VALUES
Upon successful completion,
.Nm crypt ,
.Nm crypt_r ,
.Nm crypt_rn ,
and
.Nm crypt_ra
return a pointer to a string which encodes both the hashed passphrase,
and the settings that were used to encode it.
This string is directly usable as
.Fa setting
in other calls to
.Nm crypt ,
.Nm crypt_r ,
.Nm crypt_rn ,
and
.Nm crypt_ra ,
and as
.Fa prefix
in calls to
.Nm crypt_gensalt ,
.Nm crypt_gensalt_rn ,
and
.Nm crypt_gensalt_ra .
It will be entirely printable ASCII,
and will not contain whitespace
or the characters
.Sq Li \&: ,
.Sq Li \&; ,
.Sq Li \&* ,
.Sq Li \&! ,
or
.Sq Li \&\e .
See
.Xr crypt 5
for more detail on the format of hashed passphrases.
.Pp
.Nm crypt
places its result in a static storage area,
which will be overwritten by subsequent calls to
.Nm crypt .
It is not safe to call
.Nm crypt
from multiple threads simultaneously.
.Pp
.Nm crypt_r ,
.Nm crypt_rn ,
and
.Nm crypt_ra
place their result in the
.Fa output
field of their
.Fa data
argument.
It is safe to call them from multiple threads simultaneously,
as long as a separate
.Fa data
object is used for each thread.
.Pp
Upon error,
.Nm crypt_r ,
.Nm crypt_rn ,
and
.Nm crypt_ra
write an
.Em invalid
hashed passphrase to the
.Fa output
field of their
.Fa data
argument, and
.Nm crypt
writes an invalid hash to its static storage area.
This string will be shorter than 13 characters,
will begin with a
.Sq Li \&* ,
and will not compare equal to
.Fa setting .
.Pp
Upon error,
.Nm crypt_rn
and
.Nm crypt_ra
return a null pointer.
.Nm crypt_r
and
.Nm crypt
may also return a null pointer,
or they may return a pointer to the invalid hash,
depending on how libcrypt was configured.
(The option to return the invalid hash is for compatibility
with old applications that assume that
.Nm crypt
cannot return a null pointer.
See
.Sx PORTABILITY NOTES
below.)
.Pp
All four functions set
.Va errno
when they fail.
.Sh ERRORS
.Bl -tag -width Er
.It Er EINVAL
.Fa setting
is invalid, or requests a hashing method that is not supported.
.It Er ERANGE
.Fa phrase
is too long
(more than
.Dv CRYPT_MAX_PASSPHRASE_SIZE
characters; some hashing methods may have lower limits).
.br
.Nm crypt_rn
only:
.Fa size
is too small for the hashing method requested by
.Fa setting .
.It Er ENOMEM
Failed to allocate internal scratch memory.
.br
.Nm crypt_ra
only: failed to allocate memory for
.Fa data .
.It Er ENOSYS No or Er EOPNOTSUPP
Hashing passphrases is not supported at all on this installation,
or the hashing method requested by
.Fa setting
is not supported.
These error codes are not used by this version of libcrypt,
but may be encountered on other systems.
.El
.Sh PORTABILITY NOTES
.Nm crypt
is included in POSIX, but
.Nm crypt_r ,
.Nm crypt_rn ,
and
.Nm crypt_ra
are not part of any standard.
.Pp
POSIX does not specify any hashing methods,
and does not require hashed passphrases to be portable between systems.
In practice, hashed passphrases are portable
as long as both systems support the hashing method that was used.
However, the set of supported hashing methods
varies considerably from system to system.
.Pp
The behavior of
.Nm crypt
on errors isn't well standardized.
Some implementations simply can't fail
(except by crashing the program),
others return a null pointer or a fixed string.
Most implementations don't set
.Va errno ,
but some do.
POSIX specifies returning a null pointer and setting
.Va errno ,
but it defines only one possible error,
.Er ENOSYS ,
in the case where
.Nm crypt
is not supported at all.
Some older applications are not prepared to handle null pointers
returned by
.Nm crypt .
The behavior described above for this implementation,
setting
.Va errno
and returning an invalid hashed passphrase different from
.Fa setting ,
is chosen to make these applications fail closed when an error occurs.
.Pp
Due to historical restrictions
on the export of cryptographic software from the USA,
.Nm crypt
is an optional POSIX component.
Applications should therefore be prepared for
.Nm crypt
not to be available,
or to always fail (setting
.Va errno
to
.Er ENOSYS )
at runtime.
.Pp
POSIX specifies that
.Nm crypt
is declared in
.In unistd.h ,
but only if the macro
.Dv _XOPEN_CRYPT
is defined and has a value greater than or equal to zero.
Since libcrypt does not provide
.In unistd.h ,
it declares
.Nm crypt ,
.Nm crypt_r ,
.Nm crypt_rn ,
and
.Nm crypt_ra
in
.In crypt.h
instead.
.Pp
On a minority of systems (notably recent versions of Solaris),
.Nm crypt
uses a thread-specific static storage buffer,
which makes it safe to call from multiple threads simultaneously,
but does not prevent each call within a thread
from overwriting the results of the previous one.
.Sh BUGS
Some implementations of
.Nm crypt ,
upon error,
return an invalid hash that is stored in a read-only location
or only initialized once,
which means that it is only safe to erase the buffer pointed to by the
.Nm crypt
return value if an error did not occur.
.Pp
.Vt "struct crypt_data"
may be quite large (32kB in this implementation of libcrypt;
over 128kB in some other implementations).
This is large enough that it may be unwise to allocate it on the stack.
.Pp
Some recently designed hashing methods need even more scratch memory,
but the
.Nm crypt_r
interface makes it impossible to change the size of
.Vt "struct crypt_data"
without breaking binary compatibility.
The
.Nm crypt_rn
interface could accommodate larger allocations for specific hashing methods,
but the caller of
.Nm crypt_rn
has no way of knowing how much memory to allocate.
.Nm crypt_ra
does the allocation itself,
but can only make a single call to
.Xr malloc 3 .
.Sh ATTRIBUTES
For an explanation of the terms used in this section, see
.Xr attributes 7 .
.TS
allbox;
lb lb lb
l l l.
Interface	Attribute	Value
T{
.Nm crypt
T}	Thread safety	MT-Unsafe race:crypt
T{
.Nm crypt_r ,
.Nm crypt_rn ,
.Nm crypt_ra
T}	Thread safety	MT-Safe
.TE
.sp
.Sh HISTORY
A rotor-based
.Nm crypt
function appeared in
.At v6 .
The
.Dq traditional
DES-based
.Nm crypt
first appeared in
.At v7 .
.Pp
.Nm crypt_r
originates with the GNU C Library.
There's also a
.Nm crypt_r
function on HP-UX and MKS Toolkit, but the prototypes and semantics
differ.
.Pp
.Nm crypt_rn
and
.Nm crypt_ra
originate with the Openwall project.
.Sh SEE ALSO
.Xr crypt_gensalt 3 ,
.Xr getpass 3 ,
.Xr getpwent 3 ,
.Xr shadow 3 ,
.Xr login 1 ,
.Xr passwd 1 ,
.Xr crypt 5 ,
.Xr passwd 5 ,
.Xr shadow 5 ,
.Xr pam 8


NEWS
libxcrypt NEWS -- history of user-visible changes.

Please send bug reports, questions and suggestions to
<https://github.com/besser82/libxcrypt/issues>.

Version 4.4.4
* The crypt_* functions will now all fail and set errno to ERANGE if
  their 'phrase' argument is longer than CRYPT_MAX_PASSPHRASE_SIZE
  characters (this is currently 512).  Formerly, longer passphrases
  would either be silently accepted, silently truncated, or the
  library would crash, depending on the hashing method.
* The NT hashing method no longer truncates passphrases at 128
  characters; Windows does not do this.  (The Windows login dialog
  _limits_ interactively entered passphrases to 127 characters.
  Passphrases set via the low-level API can be longer.)

Version 4.4.3
* Fix the value of SUNMD5_MAX_ROUNDS.
* Add generated C++-guards to <xcrypt.h>.
* Add --enable-obsolete-api-enosys configure option.
  If enabled, this option replaces the obsolete APIs (fcrypt,
  encrypt{,_r}, and setkey{,_r}) with stubs that set errno to
  ENOSYS and return without performing any real operations.
  This allows one to disable DEScrypt support while preserving
  POSIX compliance.
  For security reasons, the encrypt{,r} functions will also
  overwrite their data-block argument with random bits.
  The fcrypt function will always produce a failure token
  (*0 or *1), unless the library was also configured with
  --disable-failure-tokens, in which case it will always
  return NULL.

Version 4.4.2
* Add test-alg-yescrypt for improved coverage.
* Add x32 specific inline asm.

Version 4.4.1
* Change the output of the gensalt function for the NT hashing method
  ($3$) to output just its prefix ($3$) instead of a fake salt.
  The output buffer for the returned string must be at least 4 bytes
  long.
* Decrase the minimum required amount of random bytes for the NT
  hashing method to 1 byte, which is actually not used.
* The default linker flags now generate binaries with full read-only
  relocations (including full read-only re-mapping of the GOT), if
  supported by the linker and the system's native binary format.

Version 4.4.0
* Implement the crypt_preferred_method function.
  This function can be used as a convenience function to get the prefix
  of the preferred hash method.

Version 4.3.4
* --enable-hashes now supports 'fedora' as a group of hashing methods.
* Add aliases for xcrypt{,_r} and xcrypt_gensalt{,_r}.
  They were added for code compatibility with libxcrypt v3.1.1
  and earlier.
* Install the <xcrypt.h> header file, declaring the previously named
  aliases, and a symlink from libxcrypt.so to libcrypt.so, if a shared
  library is build.  For static libraries a corresponding symlink for
  the archive file will be installed.
  The installation of the compatibility files can be disabled by
  passing the '--disable-xcrypt-compat-files' flag to the configure
  script.
* Replace the prototype for the crypt_gensalt_r function with a
  declaration through a macro, so new compiled applications link
  against the identical crypt_gensalt_rn function directly.

Version 4.3.3
* Add an alias for crypt_gensalt_r.
  The function was available in older versions (v3.1.1 and earlier)
  of libxcrypt.  It has the same semantics and the same prototype as
  the crypt_gensalt_rn function.

Version 4.3.2
* Fix the gensalt function for the NT hashing method ($3$) to
  properly terminate its output string.  The output buffer for the
  returned string must be at least 30 bytes long.
* Remove the gensalt fuction for the bcrypt x variant ($2x$).
  This prefix was never intended for use when hashing new passphrases.
  The only use case originally intended was to manually edit '$2a$' to
  '$2x$' in passphrase hashes to knowingly take the risk yet enable
  users to continue to log in when upgrading systems with buggy bcrypt
  implementations to fixed versions.

Version 4.3.1
* Reduce the number of methods that can be the default for new hashes.
  We don’t want to accidentally encourage use of gost-yescrypt,
  scrypt, or the bug-compatibility bcrypt variants by people who do
  not have a specific need for them.
* sha256crypt ($5$) is not considered to be strong anymore.  Thus
  it was dropped from the STRONG and DEFAULT sets.
* The four variants of bcrypt ($2b$, $2a$, $2y$, $2x$) are now
  independently selectable at configure time.  The $2x$ variant has
  been dropped from the STRONG set, because it has a severe bug
  causing it to be easy to find collisions for some passwords.  The
  $2x$ and $2y$ variants have been dropped from the FreeBSD, NetBSD,
  OpenBSD, and Solaris compatibility sets, because those operating
  systems never supported those variants.

Version 4.3.0
* Implement the gost-yescrypt ($gy$) hashing algorithm.
* Remove all of the nonnull annotations, which are a questionable
  optimization that have caused problems in the past and may cause
  future problems on some systems.
* Rename all hash methods to match naming in John the Ripper.
* Make the configure switches --enable-obsolete-api and
  --enable-hashes process their arguments case-insensitively; for
  instance, "--enable-hashes=OpenBSD" and "--enable-hashes=openbsd"
  are now understood as synonymous.
* Fix gensalt for bigcrypt if descrypt hash method is not selected
  at compile time.
* Fix the bigcrypt hash method to reject too short settings and/or
  descrypt hashes with a truncated phrase.  Hashes with a setting
  shorter than 13 characters and phrases shorter than 8 characters
  are valid bigcrypt hashes, although they are identical with the
  output generated by descrypt under the same conditions.
  This only applies when the descrypt hash method is not selected
  at compile time.
* Implement crypt_checksalt, which can be used by portable users of
  libxcrypt to check whether the desired hash method is supported.
* Make a default prefix available whenever at least one strong hash
  is enabled.
* Fix the definition of 'CRYPT_GENSALT_IMPLEMENTS_DEFAULT_PREFIX'

crypt.c
/* High-level libcrypt interfaces.

   Copyright 2007-2017 Thorsten Kukuk and Zack Weinberg
   Copyright 2018-2019 Björn Esser

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public License
   as published by the Free Software Foundation; either version 2.1 of
   the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, see
   <https://www.gnu.org/licenses/>.  */

#include "crypt-port.h"
#include "xcrypt.h"

#include <errno.h>
#include <stdlib.h>

/* The internal storage area within struct crypt_data is used as
   follows.  We don't know what alignment the algorithm modules will
   need for their scratch data, so give it the maximum natural
   alignment.  Note that the C11 alignas() specifier can't be applied
   directly to a struct type, but it can be applied to the first field
   of a struct, which effectively forces alignment of the entire
   struct, since the first field must always have offset 0.  */
struct crypt_internal
{
  char alignas (max_align_t) alg_specific[ALG_SPECIFIC_SIZE];
};

static_assert(sizeof (struct crypt_internal) + alignof (struct crypt_internal)
              <= CRYPT_DATA_INTERNAL_SIZE,
              "crypt_data.internal is too small for crypt_internal");

/* struct crypt_data is allocated by application code and contains
   only char-typed fields, so its 'internal' field may not be
   sufficiently aligned.  */
static inline struct crypt_internal *
get_internal (struct crypt_data *data)
{
  uintptr_t internalp = (uintptr_t) data->internal;
  const uintptr_t align = alignof (struct crypt_internal);
  internalp = (internalp + align - 1) & ~(align - 1);
  return (struct crypt_internal *)internalp;
}

typedef void (*crypt_fn) (const char *phrase, size_t phr_size,
                          const char *setting, size_t set_size,
                          uint8_t *output, size_t out_size,
                          void *scratch, size_t scr_size);

typedef void (*gensalt_fn) (unsigned long count,
                            const uint8_t *rbytes, size_t nrbytes,
                            uint8_t *output, size_t output_size);

struct hashfn
{
  const char *prefix;
  size_t plen;
  crypt_fn crypt;
  gensalt_fn gensalt;
  /* The type of this field is unsigned char to ensure that it cannot
     be set larger than the size of an internal buffer in crypt_gensalt_rn.  */
  unsigned char nrbytes;
};

static const struct hashfn hash_algorithms[] =
{
  HASH_ALGORITHM_TABLE_ENTRIES
};

#if INCLUDE_descrypt || INCLUDE_bigcrypt
static int
is_des_salt_char (char c)
{
  return ((c >= 'a' && c <= 'z') ||
          (c >= 'A' && c <= 'Z') ||
          (c >= '0' && c <= '9') ||
          c == '.' || c == '/');
}
#endif

static const struct hashfn *
get_hashfn (const char *setting)
{
  const struct hashfn *h;
  for (h = hash_algorithms; h->prefix; h++)
    {
      if (h->plen > 0)
        {
          if (!strncmp (setting, h->prefix, h->plen))
            return h;
        }
#if INCLUDE_descrypt || INCLUDE_bigcrypt
      else
        {
          if (setting[0] == '\0' ||
              (is_des_salt_char (setting[0]) && is_des_salt_char (setting[1])))
            return h;
        }
#endif
    }
  return 0;
}

static void
do_crypt (const char *phrase, const char *setting, struct crypt_data *data)
{
  if (!phrase || !setting)
    {
      errno = EINVAL;
      return;
    }
  /* Do these strlen() calls before reading prefixes of either
     'phrase' or 'setting', so we get a predictable crash if they are
     not valid strings.  */
  size_t phr_size = strlen (phrase);
  size_t set_size = strlen (setting);
  if (phr_size >= CRYPT_MAX_PASSPHRASE_SIZE)
    {
      errno = ERANGE;
      return;
    }
  const struct hashfn *h = get_hashfn (setting);
  if (!h)
    {
      /* Unrecognized hash algorithm */
      errno = EINVAL;
      return;
    }

  struct crypt_internal *cint = get_internal (data);
  h->crypt (phrase, phr_size, setting, set_size,
            (unsigned char *)data->output, sizeof data->output,
            cint->alg_specific, sizeof cint->alg_specific);

  XCRYPT_SECURE_MEMSET (data->internal, sizeof data->internal);
}

#if INCLUDE_crypt_rn
char *
crypt_rn (const char *phrase, const char *setting, void *data, int size)
{
  make_failure_token (setting, data, MIN (size, CRYPT_OUTPUT_SIZE));
  if (size < 0 || (size_t)size < sizeof (struct crypt_data))
    {
      errno = ERANGE;
      return 0;
    }

  struct crypt_data *p = data;
  do_crypt (phrase, setting, p);
  return p->output[0] == '*' ? 0 : p->output;
}
SYMVER_crypt_rn;
#endif

#if INCLUDE_crypt_ra
char *
crypt_ra (const char *phrase, const char *setting, void **data, int *size)
{
  if (!*data)
    {
      *data = malloc (sizeof (struct crypt_data));
      if (!*data)
        return 0;
      *size = sizeof (struct crypt_data);
    }
  if (*size < 0 || (size_t)*size < sizeof (struct crypt_data))
    {
      void *rdata = realloc (*data, sizeof (struct crypt_data));
      if (!rdata)
        return 0;
      *data = rdata;
      *size = sizeof (struct crypt_data);
    }

  struct crypt_data *p = *data;
  make_failure_token (setting, p->output, sizeof p->output);
  do_crypt (phrase, setting, p);
  return p->output[0] == '*' ? 0 : p->output;
}
SYMVER_crypt_ra;
#endif

#if INCLUDE_crypt_r
char *
crypt_r (const char *phrase, const char *setting, struct crypt_data *data)
{
  make_failure_token (setting, data->output, sizeof data->output);
  do_crypt (phrase, setting, data);
#if ENABLE_FAILURE_TOKENS
  return data->output;
#else
  return data->output[0] == '*' ? 0 : data->output;
#endif
}
SYMVER_crypt_r;
#endif

/* For code compatibility with older versions (v3.1.1 and earlier).  */
#if INCLUDE_crypt_r && INCLUDE_xcrypt_r
strong_alias (crypt_r, xcrypt_r);
SYMVER_xcrypt_r;
#endif

#if INCLUDE_crypt_gensalt_rn
char *
crypt_gensalt_rn (const char *prefix, unsigned long count,
                  const char *rbytes, int nrbytes, char *output,
                  int output_size)
{
  make_failure_token ("", output, output_size);

  /* Individual gensalt functions will check for adequate space for
     their own breed of setting, but the shortest possible one is
     three bytes (DES two-character salt + NUL terminator) and we
     also want to rule out negative numbers early.  */
  if (output_size < 3)
    {
      errno = ERANGE;
      return 0;
    }

  /* If the prefix is 0, that means to use the current best default.
     Note that this is different from the behavior when the prefix is
     "", which selects DES.  HASH_ALGORITHM_DEFAULT is not defined when
     the current default algorithm was disabled at configure time.  */
  if (!prefix)
    {
#if defined HASH_ALGORITHM_DEFAULT
      prefix = HASH_ALGORITHM_DEFAULT;
#else
      errno = EINVAL;
      return 0;
#endif
    }

  const struct hashfn *h = get_hashfn (prefix);
  if (!h)
    {
      errno = EINVAL;
      return 0;
    }

  char internal_rbytes[UCHAR_MAX];
  /* typeof (internal_nrbytes) == typeof (h->nrbytes).  */
  unsigned char internal_nrbytes = 0;

  /* If rbytes is 0, read random bytes from the operating system if
     possible.  */
  if (!rbytes)
    {
      if (!get_random_bytes (internal_rbytes, h->nrbytes))
        return 0;

      rbytes = internal_rbytes;
      nrbytes = internal_nrbytes = h->nrbytes;
    }

  h->gensalt (count,
              (const unsigned char *)rbytes, (size_t)nrbytes,
              (unsigned char *)output, (size_t)output_size);

  if (internal_nrbytes)
    XCRYPT_SECURE_MEMSET (internal_rbytes, internal_nrbytes);

  return output[0] == '*' ? 0 : output;
}
SYMVER_crypt_gensalt_rn;
#endif

/* For code compatibility with older versions (v3.1.1 and earlier).  */
#if INCLUDE_crypt_gensalt_rn && INCLUDE_crypt_gensalt_r
strong_alias (crypt_gensalt_rn, crypt_gensalt_r);
SYMVER_crypt_gensalt_r;
#endif

/* For code compatibility with older versions (v3.1.1 and earlier).  */
#if INCLUDE_crypt_gensalt_rn && INCLUDE_xcrypt_gensalt_r
strong_alias (crypt_gensalt_rn, xcrypt_gensalt_r);
SYMVER_xcrypt_gensalt_r;
#endif

#if INCLUDE_crypt_gensalt_ra
char *
crypt_gensalt_ra (const char *prefix, unsigned long count,
                  const char *rbytes, int nrbytes)
{
  char *output = malloc (CRYPT_GENSALT_OUTPUT_SIZE);
  if (!output)
    return 0;

  char *result = crypt_gensalt_rn (prefix, count, rbytes, nrbytes, output,
                                   CRYPT_GENSALT_OUTPUT_SIZE);
  if (result == 0)
    free (output);
  return result;
}
SYMVER_crypt_gensalt_ra;
#endif

#if INCLUDE_crypt_checksalt
static_assert(CRYPT_SALT_OK == 0, "CRYPT_SALT_OK does not equal zero");

int
crypt_checksalt (const char *setting)
{
  int retval = CRYPT_SALT_INVALID;

  if (!setting)
    return retval;

  const struct hashfn *h = get_hashfn (setting);

  if (h)
    retval = CRYPT_SALT_OK;

  return retval;
}
SYMVER_crypt_checksalt;
#endif

#if INCLUDE_crypt_preferred_method
const char *
crypt_preferred_method (void)
{
#if defined HASH_ALGORITHM_DEFAULT
      return HASH_ALGORITHM_DEFAULT;
#else
      return NULL;
#endif
}
SYMVER_crypt_preferred_method;
#endif
